### ISSUE TITLE
Continue publishing items after Action in Subscribe has completed

### ISSUE THREAD

**Issue by michaelmairegger** [dotnet/reactive#404]
How do I achieve that a Subscription is executed synchronous and after the Subscription completes the last element that enters in the meantime is passed to subscription. During execution of the subscription-action no other subscription-action should be called

```sh
queue    1234567890123456789
work     1----6----1----3--- (the hyphen indicates the work)
```

The reason is because I observe on updates in a property. If the user modifies the data before the computation is completed the current recalculation should be either terminated or a new one should not be started. But after completion/termination of the running computation a new computation should be started with the last entered data.

My current code:

    this.GetPropertyAsObservable(() => AmountToProduce)
                .Sample(TimeSpan.FromSeconds(1))
                .SelectMany(ComputeAmount)
                .Subscribe();

Do you have any suggestions how to achieve my goal? Thanks

**Comment by jericsmith** [comment]
Subscribe accepts a CancellationToken in an overload.

**Comment by jericsmith** [comment]
you can also hold a reference to the subscription and call Dispose

```
var observableProperty = this.GetPropertyAsObservable(...);
var subscription = observableProperty.Subscribe(...);

subscription.Dispose();
```

**Comment by xxMUROxx** [comment]
@jericsmith That means that if I pass a CancellationToken no new action (i.e. `ComputeAmount` that I have passed to `SelectMany`) is executed until that action is completed?

**Comment by xxMUROxx** [comment]
@jericsmith I have a sample. Please run the tests one after each other (each run another test) not the whole fest-fixture.

    [TestFixture]
    [SingleThreaded]
    public class TestRx
    {
        private static bool _isRunning;
        private static int _timesConcurrentRunOccured;


        [Test]
        public async Task Fail1()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Select(DoWork)
                      .Subscribe(async i =>
                                 {
                                     await i;
                                 });

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }


        [Test]
        public async Task Fail2()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Subscribe(async i =>
                                 {
                                     await DoWork(i);
                                 });

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }


        [Test]
        public async Task Fail3()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Select(DoWork)
                      .Subscribe();

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }


        [Test]
        public async Task Fail4()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Select(async i => await DoWork(i))
                      .Subscribe();

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }



        [Test]
        public async Task Success1()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Select(i =>
                              {
                                  DoWork(i).Wait();
                                  return i;
                              })
                      .Subscribe();

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }

        [Test]
        public async Task Success2()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Subscribe(i =>
                                 {
                                     DoWork(i).Wait();
                                 });

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();
        }
        [Test]
        public async Task Success3()
        {
            var subscription = Observable.Interval(TimeSpan.FromSeconds(1))
                      .Select(DoWork)
                      .Subscribe(i =>
                                 {
                                     i.Wait();
                                 });

            await Task.Delay(TimeSpan.FromSeconds(5));
            subscription.Dispose();

        }

        [SetUp]
        public void SetUp()
        {
            _isRunning = false;
            _timesConcurrentRunOccured = 0;
        }

        [TearDown]
        public void Teardown()
        {
            Assert.That(_timesConcurrentRunOccured, Is.Zero);
        }


        private static async Task<bool> DoWork(long i)
        {
            if (_isRunning)
            {
                _timesConcurrentRunOccured++;
            }

            _isRunning = true;
            await Task.Delay(TimeSpan.FromSeconds(2));

            _isRunning = false;

            return true;
        }
    }

Why does the tests with `async` keyword fail?

**Comment by jericsmith** [comment]
I read your original question.  I had misunderstood what you were attempting.  I'm not sure Rx can handle your scenario, but I'll look into it further.

**Comment by jericsmith** [comment]
@xxMUROxx If I understand correctly, I believe your problem can be solved using scheduling.  With scheduling, it is possible to repeat action execution indefinitely.  When satisfied with the outcome, the individual action may be disposed, ending the execution process.

If that sounds like it would help you out, please attach some code mimicking your dilemma.  It will be an interesting exercise.

**Comment by xxMUROxx** [comment]
@jericsmith I will create a sample that satisfy the requirements.

**Comment by xxMUROxx** [comment]
@jericsmith A sample would be a CI-Server.

The CI-Server is set-up in such a way that a new build starts when a new commit is pushed. Normally the build takes some time. Now the desired behavior is that in the time when a build is running no new build starts. Therefore, if new commits arrive these commits are ignored until the previous commit is completed (w/o success). After finishing we might have three options:

1. No new commit
- one new commit
- multiple commits

For **one** commit we could easily start it. But if we have multiple commits it is required to start the **latest** that arrived..

as timeline we could draw:


|  | Queue |
|---|---|
| commit queue | `a.........bcdefghi..........j.........` |
| build queue  | `a------...b------i------....j------...` |
| state        | `21111111112333333311111111112111111111` |

where following holds:

| symbol | description |
| - | - |
| each single alphanumeric char | a new pushed commit (i.e. an item published to the `IObservable<T>`)
| - | work |
| . | idle |
| state number | corresponds to the number if the list above |

**Comment by jericsmith** [comment]
Sorry, I had my own issue I was working on all day.  Tell me if you can make this work.  This is directly from Reactive Programming for .NET Developers by Antonio Esposito and Michael Ciceri (pub 2016):
```
            Console.WriteLine("{0} -> Playing...", DateTime.Now);

            //a sourcing sequence 
            var sequence = Observable.Interval(TimeSpan.FromSeconds(1)).Take(5);

            var trace = sequence
                //marks each message with a timestamp 
                .Timestamp()
                //route messages into a list of timestamped messages 
                .ToList()
                //materialize the list when the sequence completes 
                //and return only the list 
                .Wait();

            //a scheduler for historical records 
            var scheduler = new HistoricalScheduler();

            Console.WriteLine("{0} -> Replaying...", DateTime.Now);

            //generate a new sequence from a collection 
            var replay = Observable.Generate(
                //the enumerator to read values from 
                trace.GetEnumerator(),
                //the condition to check until False 
                x => x.MoveNext(),
                //the item 
                x => x,
                //the item's value 
                x => x.Current.Value,
                //the item's virtual time 
                x => x.Current.Timestamp,
                //the scheduler  
                scheduler).TakeLast(1);  //I added the TakeLast(1); you said you were only interested in the last commit

            //some output 
            replay.Subscribe(x => Console.WriteLine("{0} -> {1}", scheduler.Now, x));

            //play the record 
            scheduler.Start();

            Console.ReadLine();
```

**Comment by RxDave** [comment]
This appears to be a typical feedback loop problem. These kinds of problems are a bit tricky, with several choices available. In your case, it appears that you want to drop notifications that arrive when the consumer is still processing the previous notification. These problems are also similar to the typical backpressure issues; i.e., having a single thread of execution, with a producer that is faster than the consumer. You can choose to either enqueue indefinitely, enqueue to a limit, batch by count or time, or simply drop items by either sampling or throttling.

To drop items while the previous item is still being processed, try something like this:

    IConnectableObservable<Unit> buildsCompleted = null;
    var toBuild = commits.Publish(p => p.Sample(
      Observable.Defer(() => buildsCompleted)
                .Merge(p.Take(1).Select(_ => Unit.Default))));

    buildsCompleted = toBuild.SelectMany(BuildAsync).Publish();
    buildsCompleted.Connect();

where `BuildAsync` has the following signature:

    private Task<Unit> BuildAsync(CommitData commit, CancellationToken cancel)

**Comment by RxDave** [comment]
Thinking about this particular "commits & builds" example a bit more, it also seems reasonable to me to require commits from different source URLs to be built concurrently, with some maximum concurrency specified, while ensuring that the latest commit is always built and older commits are skipped, and also while preserving the FIFO order of the commits across all source URLs.

Here's what I've come up with. It's actually a fairly small query, but you must also consider the small `CommitState` type that was defined to help with the requirement that the same commit must not be built more than once.

One issue with this query (as is fairly common with feedback-related queries) is that each commit will remain in memory rather than being purged immediately. An old commit is skipped only once the `Merge` operator gets around to subscribing to its closure observable, at which time the `CommitState` type becomes useful because it indicates that the commit has already been processed and no other later commits have arrived since, thus the triggering commit is redundant and may be skipped.

    using System;
    using System.Linq;
    using System.Reactive.Linq;
    using System.Reactive.Threading.Tasks;
    using System.Threading.Tasks;

    namespace ReactiveFeedbackLab
    {
      internal sealed class Program
      {
        private static void Main()
        {
          Console.WriteLine("Max Concurrency: " + Environment.ProcessorCount);
          Console.WriteLine();

          var commits = (from n in Observable.Interval(TimeSpan.FromSeconds(1))
                         select new Commit(n, new Uri("Source #" + (n % (Environment.ProcessorCount / 2)), UriKind.Relative)))
                        .Publish();

          var builds = (from commit in commits
                        group commit by commit.Source into commitsBySource
                        let latest = commitsBySource.Replay(1).RefCount()
                        from commit in latest
                        select Observable.Defer(() => Observable.Return(new CommitState(latest.Take(1).Wait())))
                                         .SelectMany(c => !c.Used
                                                        ? c.BuildAsync().ToObservable().Select(_ => c.Commit)
                                                        : Observable.Empty<Commit>()))
                        .Merge(maxConcurrent: Environment.ProcessorCount);

          using (commits.Subscribe(commit => Console.WriteLine("Committed: " + commit)))
          using (builds.Subscribe(commit => Console.WriteLine("Built: " + commit)))
          using (commits.Connect())
          {
            Console.ReadKey();
          }

          Console.ReadKey();
        }
      }

      internal sealed class Commit
      {
        public Commit(long number, Uri source)
        {
          Number = number;
          Source = source;
        }

        public long Number { get; }

        public Uri Source { get; }

        public Task BuildAsync()
        {
          Console.WriteLine("Building: " + this);

          // TODO: Build the commit here
          return Task.Delay(TimeSpan.FromSeconds(16.3));
        }

        public override string ToString() => Number + ", " + Source;
      }

      internal sealed class CommitState
      {
        public CommitState(Commit commit) => Commit = commit;

        public Commit Commit { get; }

        public bool Used { get; private set; }

        public Task BuildAsync()
        {
          Used = true;
          return Commit.BuildAsync();
        }
      }
    }

**Comment by RxDave** [comment]
Here's a simplified example of the above to address your original question. Note that it also suffers from the same issue above; namely, any property change that occurs while a previous action is still in progress will be enqueued by the `Concat` operator. Older changes will eventually be discarded once the `Concat` operator gets around to them, but they'll remain in memory until that time.

```csharp
using System;
using System.Linq;
using System.Reactive.Linq;
using System.Reactive.Threading.Tasks;
using System.Threading.Tasks;

namespace ReactiveFeedbackLab
{
  internal sealed class Program
  {
    private static void Main()
    {
      var propertyChanges = Observable.Interval(TimeSpan.FromSeconds(1))
                                      .Select(value => "ABC" + value);

      var latest = propertyChanges.Select((v, i) => (Value: v, Index: i)).Replay(1);
      var actions = (from pair in latest
                     select Observable.Defer(() => Observable.Return(latest.Take(1).Wait()))
                                      .SelectMany(l => pair.Index == l.Index
                                                     ? DoAsync(pair.Value).ToObservable().Select(_ => pair.Value)
                                                     : Observable.Empty<string>()))
                     .Concat();

      using (latest.Subscribe(value => Console.WriteLine("Property Changed: " + value)))
      using (actions.Subscribe(value => Console.WriteLine("Action completed: " + value)))
      using (latest.Connect())
      {
        Console.ReadKey();
      }

      Console.ReadKey();
    }

    // TODO: Implement this function
    public static Task DoAsync(string propertyValue)
    {
      Console.WriteLine("Acting on: " + propertyValue);

      return Task.Delay(TimeSpan.FromSeconds(2.3));
    }
  }
}
```

**Comment by jericsmith** [comment]
Nice.  You gotta love Rx.  This is helpful to me as well.  I’ve been trying to ramp up in order to present the benefits of Rx to management.  Examples like this are perfect.

Michael, this or something similar should satisfy your requirements, no?

**Comment by RxDave** [comment]
Ideally, I think that a new overload of the `Merge` operator would solve the remaining issue with my queries above; namely, redundant notifications should be dropped immediately rather than being enqueued, just to be dropped later.

I've implemented this at work already, but unfortunately I'm probably unable to contribute the code. IIRC, the signature was something like this:

    public static IObservable<TSource> Merge<TSource>(
        IObservable<IObservable<TSource>> sources,
        int maxConcurrent, 
        ICollection<TSource> queue)

Or it may have had a predicate parameter rather than a queue, or maybe there were overloads of both. I don't remember exactly :-)

Note that the `Concat` scenario would just be `sources.Merge(1, ...)`, so this would work for both of my examples above: the commits/builds query, and the query that addresses your original property changes problem.

**Comment by xxMUROxx** [comment]
Thanks, I will try it, and then give feedback.

**Comment by xxMUROxx** [comment]
@RxDave @jericsmith  Thanks that is exactly that I was trying to do. The issue that older changes are kept into memory is no issue for me.

**Comment by akarnokd** [comment]
Sounds like a duplicate of #401. Could you verify you are talking about the same pattern?
